In [1]:
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt

import numpy as np
import pandas as pd
from datetime import datetime
import seaborn as sns
%matplotlib inline
import wrangle as w
import os
import env

In [2]:
df = w.prep_log_data()

In [3]:
df


time                                              path  \
date                                                                     
2018-01-26  09:55:03                                                 /   
2018-01-26  09:56:02                                           java-ii   
2018-01-26  09:56:05               java-ii/object-oriented-programming   
2018-01-26  09:56:06                slides/object_oriented_programming   
2018-01-26  09:56:24                         javascript-i/conditionals   
...              ...                                               ...   
2021-04-21  16:41:51                              jquery/personal-site   
2021-04-21  16:42:02                                 jquery/mapbox-api   
2021-04-21  16:42:09                           jquery/ajax/weather-map   
2021-04-21  16:44:37  anomaly-detection/discrete-probabilistic-methods   
2021-04-21  16:44:39                                 jquery/mapbox-api   

            user_id  cohort_id             ip    id     name     slack  \
date                                                                     
2018-01-26        1        8.0   97.105.19.61   8.0  Hampton  #hampton   
2018-01-26        1        8.0   97.105.19.61   8.0  Hampton  #hampton   
2018-01-26        1        8.0   97.105.19.61   8.0  Hampton  #hampton   
2018-01-26        1        8.0   97.105.19.61   8.0  Hampton  #hampton   
2018-01-26        2       22.0   97.105.19.61  22.0    Teddy    #teddy   
...             ...        ...            ...   ...      ...       ...   
2021-04-21       64       28.0  71.150.217.33  28.0    Staff         #   
2021-04-21       64       28.0  71.150.217.33  28.0    Staff         #   
2021-04-21       64       28.0  71.150.217.33  28.0    Staff         #   
2021-04-21      744       28.0  24.160.137.86  28.0    Staff         #   
2021-04-21       64       28.0  71.150.217.33  28.0    Staff         #   

            start_date    end_date           created_at           updated_at  \
date                                                                           
2018-01-26  2015-09-22  2016-02-06  2016-06-14 19:52:26  2016-06-14 19:52:26   
2018-01-26  2015-09-22  2016-02-06  2016-06-14 19:52:26  2016-06-14 19:52:26   
2018-01-26  2015-09-22  2016-02-06  2016-06-14 19:52:26  2016-06-14 19:52:26   
2018-01-26  2015-09-22  2016-02-06  2016-06-14 19:52:26  2016-06-14 19:52:26   
2018-01-26  2018-01-08  2018-05-17  2018-01-08 13:59:10  2018-01-08 13:59:10   
...                ...         ...                  ...                  ...   
2021-04-21  2014-02-04  2014-02-04  2018-12-06 17:04:19  2018-12-06 17:04:19   
2021-04-21  2014-02-04  2014-02-04  2018-12-06 17:04:19  2018-12-06 17:04:19   
2021-04-21  2014-02-04  2014-02-04  2018-12-06 17:04:19  2018-12-06 17:04:19   
2021-04-21  2014-02-04  2014-02-04  2018-12-06 17:04:19  2018-12-06 17:04:19   
2021-04-21  2014-02-04  2014-02-04  2018-12-06 17:04:19  2018-12-06 17:04:19   

            deleted_at  program_id  
date                                
2018-01-26         NaN         1.0  
2018-01-26         NaN         1.0  
2018-01-26         NaN         1.0  
2018-01-26         NaN         1.0  
2018-01-26         NaN         2.0  
...                ...         ...  
2021-04-21         NaN         2.0  
2021-04-21         NaN         2.0  
2021-04-21         NaN         2.0  
2021-04-21         NaN         2.0  
2021-04-21         NaN         2.0  

[900223 rows x 14 columns]

In [4]:
df = df.drop(columns = 'deleted_at')

In [5]:
df.dtypes

time           object
path           object
user_id         int64
cohort_id     float64
ip             object
id            float64
name           object
slack          object
start_date     object
end_date       object
created_at     object
updated_at     object
program_id    float64
dtype: object

In [6]:
df.dropna(inplace=True)

In [7]:
df['time'] = df['time'].astype(str)
df['cohort_id'] = df['cohort_id'].astype(int)
df['id'] = df['id'].astype(int)
df['start_date'] = pd.to_datetime(df['start_date'])
df['end_date'] = pd.to_datetime(df['end_date'])
df['created_at'] = pd.to_datetime(df['created_at'])
df['updated_at'] = pd.to_datetime(df['updated_at'])
df['program_id'] = df['program_id'].astype(int)

In [8]:
df.program_id.value_counts()

2    713365
3    103411
1     30548
4         5
Name: program_id, dtype: int64

In [9]:
df.isnull().sum()

time          0
path          0
user_id       0
cohort_id     0
ip            0
id            0
name          0
slack         0
start_date    0
end_date      0
created_at    0
updated_at    0
program_id    0
dtype: int64

In [10]:
df.shape

(847329, 13)

# 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

## Program 1

In [11]:
lessons = df[["program_id", "cohort_id", "path"]]
program_1_lessons = lessons[lessons["program_id"] == 1]

In [12]:
program_1_cohort_lesson_traffic = program_1_lessons.groupby(["cohort_id", "path"]).size().reset_index(name="traffic")

In [ ]:
lesson_traffic_total = program_1_cohort_lesson_traffic.groupby("path")["traffic"].sum().reset_index()


top_lessons = lesson_traffic_total.nlargest(4, "traffic")

print("Top 4 most visited lessons:")
for index, lesson in top_lessons.iterrows():
    print("Lesson Path:", lesson["path"])
    print("Total Traffic:", lesson["traffic"])
    print("---------------------------")


In [13]:
lesson_traffic_total = program_1_cohort_lesson_traffic.groupby("path")["traffic"].sum().reset_index()


top_lessons = lesson_traffic_total.nlargest(4, "traffic")

print("Top 4 most visited lessons:")
for index, lesson in top_lessons.iterrows():
    print("Lesson Path:", lesson["path"])
    print("Total Traffic:", lesson["traffic"])
    print("---------------------------")


Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 1681
---------------------------
Lesson Path: index.html
Total Traffic: 1011
---------------------------
Lesson Path: javascript-i
Total Traffic: 736
---------------------------
Lesson Path: html-css
Total Traffic: 542
---------------------------


## Program 2

In [14]:
lessons = df[["program_id", "cohort_id", "path"]]
program_2_lessons = lessons[lessons["program_id"] == 2]

In [15]:
program_2_cohort_lesson_traffic = program_2_lessons.groupby(["cohort_id", "path"]).size().reset_index(name="traffic")

In [16]:
lesson_traffic_total = program_1_cohort_lesson_traffic.groupby("path")["traffic"].sum().reset_index()


top_lessons = lesson_traffic_total.nlargest(4, "traffic")

print("Top 4 most visited lessons:")
for index, lesson in top_lessons.iterrows():
    print("Lesson Path:", lesson["path"])
    print("Total Traffic:", lesson["traffic"])
    print("---------------------------")


Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 1681
---------------------------
Lesson Path: index.html
Total Traffic: 1011
---------------------------
Lesson Path: javascript-i
Total Traffic: 736
---------------------------
Lesson Path: html-css
Total Traffic: 542
---------------------------


# Program 3

In [17]:
lessons = df[["program_id", "cohort_id", "path"]]
program_3_lessons = lessons[lessons["program_id"] == 3]

In [18]:
program_3_cohort_lesson_traffic = program_3_lessons.groupby(["cohort_id", "path"]).size().reset_index(name="traffic")

In [19]:
lesson_traffic_total = program_3_cohort_lesson_traffic.groupby("path")["traffic"].sum().reset_index()


top_lessons = lesson_traffic_total.nlargest(4, "traffic")

print("Top 4 most visited lessons:")
for index, lesson in top_lessons.iterrows():
    print("Lesson Path:", lesson["path"])
    print("Total Traffic:", lesson["traffic"])
    print("---------------------------")


Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 8358
---------------------------
Lesson Path: search/search_index.json
Total Traffic: 2203
---------------------------
Lesson Path: classification/overview
Total Traffic: 1785
---------------------------
Lesson Path: 1-fundamentals/modern-data-scientist.jpg
Total Traffic: 1655
---------------------------


## Program 4

In [20]:
lessons = df[["program_id", "cohort_id", "path"]]
program_4_lessons = lessons[lessons["program_id"] == 4]

In [21]:
program_4_cohort_lesson_traffic = program_4_lessons.groupby(["cohort_id", "path"]).size().reset_index(name="traffic")

In [22]:
program_4_cohort_lesson_traffic

cohort_id                                         path  traffic
0          9                                            /        1
1          9                             content/html-css        2
2          9  content/html-css/gitbook/images/favicon.ico        1
3          9           content/html-css/introduction.html        1

In [23]:
lesson_traffic_total = program_4_cohort_lesson_traffic.groupby("path")["traffic"].sum().reset_index()


top_lessons = lesson_traffic_total.nlargest(10, "traffic")

print("Top 4 most visited lessons:")
for index, lesson in top_lessons.iterrows():
    print("Lesson Path:", lesson["path"])
    print("Total Traffic:", lesson["traffic"])
    print("---------------------------")


Top 4 most visited lessons:
Lesson Path: content/html-css
Total Traffic: 2
---------------------------
Lesson Path: /
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/gitbook/images/favicon.ico
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/introduction.html
Total Traffic: 1
---------------------------


### In conclusion here are the most visited lessons across cohorts per program:

- program 1: javascript-i
- program 2: javascript-i
- program 3: classification/overview
- program 4: content/html-css

# 7. And Which lessons are least accessed?

In [24]:
pd.set_option('display.max_colwidth', None)

all_programs_data = [program_1_cohort_lesson_traffic, program_2_cohort_lesson_traffic, program_3_cohort_lesson_traffic]  # Add more dataframes if needed
combined_data = pd.concat(all_programs_data)

lesson_traffic_total = combined_data.groupby("path")["traffic"].sum().reset_index()

pattern = "^[^.]+/[^.]+$"


lessons_with_traffic_equal_to_1 = lesson_traffic_total[
    (lesson_traffic_total["traffic"] == 1) &
    (lesson_traffic_total["path"].str.contains(".+/.+")) &
    (lesson_traffic_total["path"].str.match(pattern))
]

In [26]:
lessons_with_traffic_equal_to_1

path  traffic
63           10-nlp/2-regular-expressions        1
64                       10-nlp/3-acquire        1
68                       10-nlp/exercises        1
99    11-distributed-ml/3-getting-started        1
100           11-distributed-ml/4-acquire        1
...                                   ...      ...
2203                         users/1/edit        1
2204                            users/167        1
2205                         users/2/edit        1
2206                            users/913        1
2215                        web-design/ux        1

[185 rows x 2 columns]

In [27]:
lessons_with_traffic_equal_to_1.shape

(185, 2)

## Looking through the lessons with the least traffic which was 1, there were lessons that no longer exist in the current curriculum and had testing titles such as asdfasdf as the lesson name. This tells me that there may at some point have been testing website pages. I would need a current website paths to ensure I am not delivering lessons that were "least accessed" that were never technically a lesson to began with. Using the current data, here are the 5 least visited lessons


- nlp/2-regular-expressions
- nlp/3-acquire
- nlp/exercises
- distributed-ml/3-getting-started
- distributed-ml/4-acquire

*** There are 185 more lessons with a traffic of 1, this is just 5 ***